File: DiseaseMutationsApp/HGVS.cs
using System.Text.RegularExpressions;

namespace DiseaseMutationsApp;

public partial class HGVS
{
    private string Code { get; set; }
    public string Accession { get; private set; }
    public string Type { get; private set; }
    public (int start, int end) Position { get; private set; }
    
    public enum MutationType
    {
        Substitution,
        Deletion,
        Insertion,
        Duplication,
        Inversion,
        DeletionInsertion,
        Repeat,
        NoChange,
        Conversion,
        Translocation,
        LargeUnknown
    }
    public MutationType 
        Mutation { get; private set; }
    
    public string Reference { get; private set; }
    public string Alternate { get; private set; }
    public int Count { get; private set; }
    
    
    public HGVS(string code)
    {
        Code = code;
        Parse();
    }

    private void Parse()
    {
        var parts = Code.Split(':');
        Accession = parts[0];
        
        
        var typeAndRest = parts[1];
        
        Type = typeAndRest[0].ToString();
        var positionAndMethod = typeAndRest[2..];
        
        //use regex to extract number or 2 numbers with underscore
        var positionMatch = PositionRange().Match(positionAndMethod);
        if (!positionMatch.Success) throw new Exception("Invalid HGVS format: Position not found");
        var positions = positionMatch.Value.Split("_");
        Position = 
            positions.Length == 2 ? 
                (int.Parse(positions[0]), int.Parse(positions[1])) :
                (int.Parse(positions[0]), int.Parse(positions[0]));
        
        var method = positionAndMethod[positionMatch.Length..];
        ParseMethod(method);
    }

    private void ParseMethod(string method)
    {
        if(string.IsNullOrEmpty(method) || method == "=")
        {
            Mutation = MutationType.NoChange;
        }
        else if (method.Contains('>'))
        {
            Mutation = MutationType.Substitution;
            var refs = method.Split('>');
            Reference = refs[0];
            Alternate = refs[1];
        }
        else if (method.StartsWith("delins"))
        {
            Mutation = MutationType.DeletionInsertion;
            Alternate = method[6..];
        }
        else if (method.StartsWith("del"))
        {
            Mutation = MutationType.Deletion;
        }
        else if (method.StartsWith("ins"))
        {
            Mutation = MutationType.Insertion;
            Alternate = method[3..];
        }
        else if (method.StartsWith("dup"))
        {
            Mutation = MutationType.Duplication;
        }
        else if (method.StartsWith("inv"))
        {
            Mutation = MutationType.Inversion;
        }
        else if (method.Contains('[') && method.Contains(']'))
        {
            Mutation = MutationType.Repeat;
            var sequence = method[..method.IndexOf('[')];
            var repeatCountStr = method[(method.IndexOf('[') + 1)..method.IndexOf(']')];
            var repeatCount = int.Parse(repeatCountStr);
            Alternate = sequence;
            Count = repeatCount;
        }
        else
        {
            throw new Exception("Invalid HGVS format: Unknown mutation method");
        }
    }

    [GeneratedRegex(@"^(\d+)(_(\d+))?")]
    private static partial Regex PositionRange();
}

File: gRNA/HGVS.fs
module gRNA.HGVS

open System
open System.Text.RegularExpressions

type MutationType =
    | Substitution
    | Deletion
    | Insertion
    | Duplication
    | Inversion
    | DeletionInsertion
    | Repeat
    | NoChange
    | Conversion
    | Translocation
    | LargeUnknown

type HGVS(code: string) =
    let positionRangeRegex = Regex(@"^(\d+)(_(\d+))?")
    
    let parseMethod (method: string) =
        if String.IsNullOrEmpty(method) || method = "=" then
            (MutationType.NoChange, "", "", 0)
        elif method.Contains('>') then
            let refs = method.Split('>')
            (MutationType.Substitution, refs.[0], refs.[1], 0)
        elif method.StartsWith("delins") then
            (MutationType.DeletionInsertion, "", method.[6..], 0)
        elif method.StartsWith("del") then
            (MutationType.Deletion, "", "", 0)
        elif method.StartsWith("ins") then
            (MutationType.Insertion, "", method.[3..], 0)
        elif method.StartsWith("dup") then
            (MutationType.Duplication, "", "", 0)
        elif method.StartsWith("inv") then
            (MutationType.Inversion, "", "", 0)
        elif method.Contains('[') && method.Contains(']') then
            let sequence = method.[..method.IndexOf('[') - 1]
            let repeatCountStr = method.[method.IndexOf('[') + 1..method.IndexOf(']') - 1]
            let repeatCount = int repeatCountStr
            (MutationType.Repeat, "", sequence, repeatCount)
        else
            raise (Exception("Invalid HGVS format: Unknown mutation method"))
    
    let parse () =
        let parts = code.Split(':')
        let accession = parts.[0]
        
        let typeAndRest = parts.[1]
        let typeValue = string typeAndRest.[0]
        let positionAndMethod = typeAndRest.[2..]
        
        let positionMatch = positionRangeRegex.Match(positionAndMethod)
        if not positionMatch.Success then
            raise (Exception("Invalid HGVS format: Position not found"))
        
        let positions = positionMatch.Value.Split('_')
        let position =
            if positions.Length = 2 then
                (int positions.[0], int positions.[1])
            else
                (int positions.[0], int positions.[0])
        
        let method = positionAndMethod.[positionMatch.Length..]
        let (mutation, reference, alternate, count) = parseMethod method
        
        (accession, typeValue, position, mutation, reference, alternate, count)
    
    let accession, typeValue, position, mutation, reference, alternate, count = parse()
    
    member _.Code = code
    member _.Accession = accession
    member _.Type = typeValue
    member _.Position = position
    member _.Mutation = mutation
    member _.Reference = reference
    member _.Alternate = alternate
    member _.Count = count
    
    member _.GetMutationLength() =
        let start, end' = position
        end' - start + 1